src/plone/restapi/services/roles/get.py

```diff
@@ -1,19 +1,44 @@
+from AccessControl import getSecurityManager
 from Acquisition import aq_inner
 from plone.restapi.services import Service
+from Products.CMFCore.permissions import ManagePortal
 from Products.CMFCore.utils import getToolByName
 from zope.i18n import translate
 
 
 class RolesGet(Service):
+    @property
+    def is_zope_manager(self):
+        return getSecurityManager().checkPermission(ManagePortal, self.context)
+
+    def get_user_roles(self):
+        membership = getToolByName(self.context, "portal_membership")
+        member = membership.getAuthenticatedMember()
+        return member.getRoles()
+
+    def can_assign(self, is_zope_manager, _id):
+        if is_zope_manager:
+            return True
+        return _id != "Manager"
+
+    def can_assign_add(self, is_zope_manager, user_roles, _id):
+        if is_zope_manager:
+            return True
+        return _id in user_roles
+
     def reply(self):
         pmemb = getToolByName(aq_inner(self.context), "portal_membership")
         roles = [r for r in pmemb.getPortalRoles() if r != "Owner"]
+        is_zope_manager = self.is_zope_manager
+        user_roles = self.get_user_roles()
         return [
             {
                 "@type": "role",
                 "@id": f"{self.context.absolute_url()}/@roles/{r}",
                 "id": r,
                 "title": translate(r, context=self.request, domain="plone"),
+                "can_assign": self.can_assign(is_zope_manager, r),
+                "can_assign_add": self.can_assign_add(is_zope_manager, user_roles, r),
```

Comment:
@wesleybl It doesn't make sense to me either, and it feels like an inconsistency that is more likely there by accident than by design. I think we can decide to do something different in the REST API for the sake of usability, and always allow assigning any role (except Manager when being assigned by non-Managers). I would prefer to do that than add both `can_assign` and `can_assign_add`, since it's hard to explain and locks us into supporting that in the API indefinitely.

Reply:
@davisagli It won't be easy to do this. See my comment at: https://github.com/plone/volto/issues/3917#issuecomment-1730277651

Reply:
@wesleybl Okay, I think I understand: CMFCore's RegistrationTool.addMember enforces this, and the plone.restapi user add service currently uses it.

It doesn't look terribly hard to work around to me: we could make the user add service first add a member without specifying additional roles, and then separately add the roles the same way that it is done when updating an existing user.

I'm okay with calling this out of scope for the current PR, but I still don't really want to add `can_assign_add` to the API, since in the long run I think we should fix it so that there's no difference in which roles can be specified when adding vs editing a user.

Reply:
@davisagli I removed the key `can_assign_add`